File: src/worktree.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

export interface WorktreeInfo {
  path: string;
  branch: string;
  isDetached: boolean;
  status: string;
}

export async function listWorktrees(cwd: string): Promise<string> {
  const { stdout } = await execFileAsync("git", ["worktree", "list"], { cwd });
  return formatWorktrees(parseWorktreeList(stdout));
}

export async function activeWorktree(cwd: string): Promise<string> {
  const { stdout } = await execFileAsync("git", ["worktree", "list"], { cwd });
  const worktrees = parseWorktreeList(stdout);
  const top = (await execFileAsync("git", ["rev-parse", "--show-toplevel"], { cwd })).stdout.trim();
  return worktrees.find((worktree) => worktree.path === top)?.path ?? top;
}

export async function createWorktree(cwd: string, target: string, branch?: string): Promise<string> {
  const args = ["worktree", "add", target];
  if (branch) args.push(branch);
  const { stdout, stderr } = await execFileAsync("git", args, { cwd });
  return [stdout, stderr].filter(Boolean).join("\n").trim();
}

export async function removeWorktree(cwd: string, target: string): Promise<string> {
  const { stdout, stderr } = await execFileAsync("git", ["worktree", "remove", target], { cwd });
  return [stdout, stderr].filter(Boolean).join("\n").trim();
}

export function parseWorktreeList(output: string): WorktreeInfo[] {
  const worktrees: WorktreeInfo[] = [];
  const lines = output.split(/\r?\n/);
  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index];
    const match = /^(\S+)\s+([a-f0-9]+)\s+(?:\[([^\]]*)\])?/.exec(line);
    if (!match) continue;
    const branch = match[3] ?? match[2];
    const isDetached = !match[3] || /^[a-f0-9]{40}$/.test(match[3]);
    let status = "";
    if (lines[index + 1]?.startsWith("# ")) {
      status = lines[index + 1].slice(2);
      index += 1;
    }
    worktrees.push({ path: match[1], branch, isDetached, status });
  }
  return worktrees;
}

export function formatWorktrees(worktrees: WorktreeInfo[]): string {
  return worktrees.map((worktree) => {
    const branch = worktree.isDetached ? `(detached ${worktree.branch})` : `[${worktree.branch}]`;
    return [worktree.path, branch, worktree.status].filter(Boolean).join("\t");
  }).join("\n");
}

File: core-go/internal/git/worktree.go
package git

import (
	"os"
	"os/exec"
	"regexp"
	"strings"
)

// WorktreeInfo contains information about a git worktree
type WorktreeInfo struct {
	Path     string
	Branch   string
	IsDetached bool
	Status   string
}

// ListWorktrees executes `git worktree list` and parses the output
// to return a slice of WorktreeInfo structures.
func ListWorktrees() ([]WorktreeInfo, error) {
	cmd := exec.Command("git", "worktree", "list")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var worktrees []WorktreeInfo
	lines := strings.Split(string(output), "\n")

	// Regex pattern to match worktree lines
	// Format: /path/to/worktree  commit-hash [branch-name]
	// or: /path/to/worktree  commit-hash (no branch)
	worktreePattern := regexp.MustCompile(`^(\S+)\s+([a-f0-9]+)\s+\[([^\]]*)\]`)

	for i := 0; i < len(lines); i++ {
		line := lines[i]
		matches := worktreePattern.FindStringSubmatch(line)
		if matches != nil {
			path := matches[1]
			commitHash := matches[2]
			branchName := matches[3]

			info := WorktreeInfo{
				Path: path,
			}

			// Check if detached (branch name is empty or looks like a commit hash)
			if branchName == "" || (len(branchName) == 40 && regexp.MustCompile(`^[a-f0-9]+$`).MatchString(branchName)) {
				info.IsDetached = true
				info.Branch = commitHash
			} else {
				info.IsDetached = false
				info.Branch = branchName
			}

			// Check if next line is a status line (starts with "# ")
			if i+1 < len(lines) && strings.HasPrefix(lines[i+1], "# ") {
				info.Status = strings.TrimPrefix(lines[i+1], "# ")
				i++ // Skip the status line
			}

			worktrees = append(worktrees, info)
		}
	}

	return worktrees, nil
}

// CreateWorktree executes `git worktree add <path> <branch>` to create a new worktree.
func CreateWorktree(path, branch string) error {
	cmd := exec.Command("git", "worktree", "add", path, branch)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return err
	}
	_ = output // Output can be logged if needed
	return nil
}

// RemoveWorktree executes `git worktree remove <path>` to remove a worktree.
func RemoveWorktree(path string) error {
	cmd := exec.Command("git", "worktree", "remove", path)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return err
	}
	_ = output // Output can be logged if needed
	return nil
}

// GetActiveWorktree returns the current worktree path by comparing
// the current working directory with the paths from `git worktree list`.
// If no matching worktree is found, it returns the main repository path.
func GetActiveWorktree() (string, error) {
	// Get current working directory
	cwd, err := os.Getwd()
	if err != nil {
		return "", err
	}

	// Get all worktrees
	worktrees, err := ListWorktrees()
	if err != nil {
		return "", err
	}

	// Compare CWD with worktree paths
	for _, wt := range worktrees {
		if wt.Path == cwd {
			return wt.Path, nil
		}
	}

	// If no match found, return the main repository path
	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(output)), nil
}
